w3af/plugins/audit/hpp.py

```diff
@@ -0,0 +1,263 @@
+from lxml import etree
+from w3af.core.controllers.exceptions import BaseFrameworkException
+from w3af.core.data.parsers.url import URL
+
+REFLECT_TOKEN = 'w3afHPPReflectToken'
+FORGE_TOKEN = 'w3afHPPForgeToken'
+EQUALITY_THRESHOLD = 0.95
+# paramater precedence types
+LAST = 'last occurance'
+FIRST = 'first occurance'
+CONCAT = 'concatenation'
+
+
+def get_forgeable(reflection, prec):
+    #this function returns list of querystring parameters which may be
+    #forged provided precedence is prec
+    r = []
+    param, url = reflection
+    qs = url.get_querystring()
+
+    passed = 0
+    for p in qs:
+        if p == param:
+            passed = 1
+            continue
+        if passed == 0 and prec == LAST:
+            r += [p]
+        if passed == 1 and prec == FIRST:
+            r += [p]
+    return r
+
+
+def find_reflection(response, precedence, patterns):
+    #finds reflection of a string or two concatenated
+    # strings in web page links or input parameters
+    if len(patterns) > 1:
+        pattern1, pattern2 = patterns
+    else:
+        pattern1, pattern2 = patterns[0], None
+
+    try:
+        parser = parser_cache.dpc.get_document_parser_for(response)
+    except BaseFrameworkException:
+        return
+
+    pattern1 = unicode(pattern1)
+    result = []
+    #find reflection in links
+    for references in parser.get_references():
+        for ref in references:
+            if not ref:
+                continue
+            qs = ref.get_querystring()
+            for p in qs:
+                if precedence == LAST and len(qs[p]) == 2:
+                    rv = qs[p][1]
+                else:
+                    rv = qs[p][0]
+                #looking for single reflection if only one pattern was
+                # entered
+                if rv == pattern1 and not pattern2:
+                    result += [(p, ref)]
+
+                if not pattern2:
+                    continue
+                #or for concatenation of both were
+                if rv[:len(pattern1)] == pattern1 and \
+                   rv[-len(pattern2):] == pattern2:
+                    result += [p, ref]
+
+    return result
+
+
+class hpp(AuditPlugin):
+    """
+    Identify client-side http parameter pollution vulnerabilities.
+    """
+
+    def __init__(self):
+        AuditPlugin.__init__(self)
+        self.parser = etree.HTMLParser(remove_blank_text=True, remove_comments=True)
+
+    def audit(self, freq, orig_response):
+        """
+        Tests an URL for HPP vulnerabilities.
+        :author: Dmitry Roshchin (nixwizard@gmail.com)
+        """
+
+        if not orig_response.is_text_or_html():
+            return
+
+        if not orig_response.get_code() == 200:
+            return
+
+        method = freq.get_method()
+
+        if method == 'GET':
+            qs = freq.get_querystring()
+
+        precedence = None
+        #finding reflected params if any
+        for param in qs:
+            value = qs[param][0]
```

Comment:
There might be more than one query string parameter with the same name, for example:

`http://foo.com/?q=abc&q=def`

By using `qs[param][0]` you're only testing for HPP in the first `q`.

On a related note, what would happen if the app has two `q` params "by default" (without any fuzzing), will the plugin go crazy?